From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Fri, 13 Sep 2024 06:10:38 +0300
Subject: [PATCH] Handle special nbt cases for Item Modifiers API

The following cases are currently handled:
- Items in bundles
- Items in containers (shulkers, chests, etc.)
- Use remainders

diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
index b38d48dea5b63395a9df9ec519e5a2d607bb4193..637919f7027057c570eb33b7b56bf98b82bf8eac 100644
--- a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
@@ -1,6 +1,10 @@
 package me.sosedik.kiterino.modifier.item;
 
 import com.mojang.datafixers.util.Pair;
+import io.papermc.paper.datacomponent.DataComponentTypes;
+import io.papermc.paper.datacomponent.item.BlockItemDataProperties;
+import io.papermc.paper.datacomponent.item.BundleContents;
+import io.papermc.paper.datacomponent.item.UseRemainder;
 import me.sosedik.kiterino.KiterinoConfig;
 import me.sosedik.kiterino.inventory.InventorySlotHelper;
 import me.sosedik.kiterino.modifier.item.context.ItemModifierContext;
@@ -66,6 +70,9 @@ import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
 import org.bukkit.NamespacedKey;
 import org.bukkit.World;
+import org.bukkit.block.BlockState;
+import org.bukkit.block.BlockType;
+import org.bukkit.block.data.BlockData;
 import org.bukkit.craftbukkit.CraftEquipmentSlot;
 import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
@@ -126,6 +133,83 @@ public class ItemModifiersHandlerImpl extends ItemModifiersHandler {
 			modified = true;
 		}
 
+		// Special nbt cases
+		org.bukkit.inventory.ItemStack item = contextBox.getItem();
+		if (!item.isEmpty()) {
+			// Bundles
+			if (item.hasData(DataComponentTypes.BUNDLE_CONTENTS)) {
+				BundleContents bundleContents = item.getData(DataComponentTypes.BUNDLE_CONTENTS);
+				assert bundleContents != null;
+				List<org.bukkit.inventory.ItemStack> items = new ArrayList<>(bundleContents.contents());
+				for (int i = 0; i < items.size(); i++) {
+					org.bukkit.inventory.ItemStack bundleItem = items.get(i);
+					org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, bundleItem));
+					if (newItem != null) {
+						items.set(i, bundleItem);
+						modified = true;
+					}
+				}
+				if (modified) {
+					item.setData(DataComponentTypes.BUNDLE_CONTENTS, BundleContents.bundleContents(items));
+				}
+			}
+			// Block storages
+			if (false && item.getType().isBlock() && item.hasData(DataComponentTypes.BLOCK_DATA)) { // TODO not implemented yet
+				BlockItemDataProperties blockItemDataProperties = item.getData(DataComponentTypes.BLOCK_DATA);
+				BlockType blockType = item.getType().asBlockType();
+				assert blockItemDataProperties != null;
+				assert blockType != null;
+				BlockData blockData = blockItemDataProperties.createBlockData(blockType);
+				BlockState blockState = blockData.createBlockState();
+				if (blockState instanceof org.bukkit.block.Container container) {
+					var inventory = container.getInventory();
+					for (int i = 0; i < inventory.getSize(); i++) {
+						org.bukkit.inventory.ItemStack containerItem = inventory.getItem(i);
+						if (containerItem == null) continue;
+
+						org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, containerItem));
+						if (newItem != null) {
+							inventory.setItem(i, containerItem);
+							modified = true;
+						}
+					}
+					if (modified) {
+						item.setData(DataComponentTypes.BLOCK_DATA, BlockItemDataProperties.blockItemStateProperties().build()); // TODO not implemented yet
+					}
+				}
+			}
+			if (item.getItemMeta() instanceof org.bukkit.inventory.meta.BlockStateMeta meta && meta.hasBlockState()) {
+				org.bukkit.block.BlockState blockState = meta.getBlockState();
+				if (blockState instanceof org.bukkit.block.Container container) {
+					var inventory = container.getInventory();
+					for (int i = 0; i < inventory.getSize(); i++) {
+						org.bukkit.inventory.ItemStack containerItem = inventory.getItem(i);
+						if (containerItem == null) continue;
+
+						org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, containerItem));
+						if (newItem != null) {
+							inventory.setItem(i, containerItem);
+							modified = true;
+						}
+					}
+					if (modified) {
+						meta.setBlockState(blockState);
+						item.setItemMeta(meta);
+					}
+				}
+			}
+			// Use remainder
+			if (item.hasData(DataComponentTypes.USE_REMAINDER)) {
+				UseRemainder useRemainder = item.getData(DataComponentTypes.USE_REMAINDER);
+				assert useRemainder != null;
+				org.bukkit.inventory.ItemStack newItem = ItemModifier.modifyItem(new ItemContextBox(contextBox.getViewer(), contextBox.getLocale(), ItemModifierContextType.EMPTY_NO_LORE, ItemModifierContext.EMPTY, useRemainder.transformInto()));
+				if (newItem != null) {
+					modified = true;
+					item.setData(DataComponentTypes.USE_REMAINDER, UseRemainder.useRemainder(newItem));
+				}
+			}
+		}
+
 		return modified ? contextBox.getItem() : null;
 	}
 
